enhancements/microshift/microshift-apiserver-custom-certs.md

```diff
@@ -0,0 +1,228 @@
+
+
+> Anytime the document mentions API server, it refers to kube API server.
+
+
+## Motivation
+Currently, users of Microshift cannot provide their own generated certificates,
+Some customers have very strict requirements regarding TLS certs. There are frequently intermediate proxies which allow only TLS traffic with certs that are signed by organization owned CAs.
+
+### User Stories
+* As a MicroShift administrator, I want to be able to add  organization generated certificates. 
+  - each certificate may contain:
+    - Single Common Name containing The API server DNS/IPAddress or a wildcard entry (wildcard certificate).
+    - Multiple Subject Alternative Names (SAN) containing the API server DNS/IPAddress.
+
+* As a MicroShift administrator, I want to provide additional DNS names for each certificate (in the Microshift config).
+
+### Goals
+* Allow MicroShift administrator to configure Microshift with externally generated certificates and
+domain names.
+
+### Non-Goals
+Automatic certificate rotation and integration with 3rd party cert issuing services.
+
+## Proposal
+Proposal suggests to provide Administrator means adding their own Certificates using microshift configuration file (/etc/microshift/config.yaml).
+
+A new `apiServer` level section will be added to the configuration called `namedCertificates`:
+
+```yaml
+apiServer:
+  InvalidCertificatesPolicy: ignore
+  namedCertificates:
+  - certPath: ~/certs/api_fqdn_1.crt
+    keyPath:  ~/certs/api_fqdn_1.key
+  - certPath: ~/certs/api_fqdn_2.crt
+    keyPath:  ~/certs/api_fqdn_2.key
+    names:
+    - api_fqdn_1     
+    - *.apps.external.com
+
+```  
+For each provided certificate, the following configuration is proposed:
+1. `names` - optional list of explicit DNS names (leading wildcards allowed).
+   If no names are provided, the implicit names will be extracted from the certificates.
+1. `certPath` -  certificate full path.
+1. `keyPath` -   certificate key full path.
+
+Certificate files will be read from their configured location by the Microshift service,
+each certification file will be validated (see validation rules).
+
+because we dont want to disrupt the internal API communication and make sure the internal certificate will be automaticly renewed,
+Those certificate will extend the default  API server [external](https://github.com/openshift/microshift/blob/main/pkg/controllers/kube-apiserver.go#L194) certificate configuration.
+
+InvalidNamedCertificatesPolicy parameter option is for security conscious users which would configure microshift to not start when the provided certificates are invalid (e.g. when they've been revoked, expired, etc.).
+this enum field may contain the following values:
+1. allow - allow to start Microshift with invalid certificates .
+1. ignore (default) - ignore invalid certificates and allow to start Microshift (those certs will not be used).
```

Comment:
setting it to `refuse`  can cause a service disruption,  Microshift node with custom certs  been running for a long time and certificates become expired , so it wont start after reboot.

Reply:
> Are we checking dates for custom certificates too? MicroShift does it with its own certs, is this something we are considering doing for warnings/errors?
> 
> MicroShift logs errors with the rotation dates, should we add a warning if any custom certs are approaching the `NotAfter` date?
> 
> ```shell
> Mar 11 14:48:52 microshift-dev microshift[1428]: ??? E0311 14:48:52.717936    1428 certchains.go:120] [kube-apiserver-external-signer kube-external-serving] rotate at: 2024-10-08 16:32:26 +0000 UTC
> ```
yes, we want to add this check, to determine if the certificate is considered as invalid.

Reply:
as agreed in slack thread,  such  option can be a cause of   bricked  device (in  case is something wrong with the certificates),
we  check  if the Certificate can be read and parsed, in case of a  bad certificate it will be skipped and not passed on to the api-server.